src/api.py

```diff
@@ -90,18 +102,55 @@ def _search_posts_query(username: str, search_content: str) -> Optional[dict]:
                 },
             ],
         }
-
-    # avoid an empty $or clause which will cause an error
-    if not username_query and not content_query:
+    else:
         return None
 
-    if username_query and not content_query:
-        return username_query
+    return content_query
 
-    if content_query and not username_query:
-        return content_query
 
-    return {"$and": [username_query, content_query]}
+def _gather_query_parts(*parts: Optional[dict]) -> Optional[list]:
```

Comment:
Variadic function arguments actually come in as a tuple:

```
>>> def fn(*args):
...     print(args)
... 
>>> fn(1, 2)
(1, 2)
>>> fn(1, 2, 3)
(1, 2, 3)
>>> fn()
()
```

Even when the function is called with zero arguments the tuple is still there, so I think the `Optional[dict]` here can be changed to just a `tuple`. I'm not sure exactly how to type it but it seems like from [this SO answer](https://stackoverflow.com/questions/67920245/can-the-unpacking-operator-be-typed-in-python-or-any-other-variadic-args-fu) it could be the following:

```suggestion
def _gather_query_parts(*parts: tuple[Optional[dict]*]) -> Optional[list]:
```

But seriously don't quote me on that lol

Reply:
I believe with the positional arguments, the type definition is the type of each individual argument (since the group of args is expected to be a tuple in all cases as you describe). Additionally this should only be `Optional[dict]` if it's possible that `None` might be explicitly handed to the function (e.g. `_gather_query_parts({1: 2}, None, {3: 4})`). If this is expected to receive a variable number of dicts, then the typing here should just be `*parts: dict`.

Hope that makes sense!

Edit: It looks like a lot of `None`s are being returned, so I bet it should be `Optional[dict]` on further thought 😄

Reply:
Oh really, that's interesting. Disregard my comment then!

Reply:
Yes, there are None's returned to gather_query_parts. The functions job is to remove None's for the most part!